document/core/valid/instructions.rst

```diff
@@ -232,6 +232,439 @@ Reference Instructions
+
+.. math::
+   \frac{
+     C \vdashreftype \X{rt} \ok
+     \qquad
+     C \vdashreftypematch \X{rt} \matchesreftype \X{rt}'
+   }{
+     C \vdashinstr \REFCAST~\X{rt} : [\X{rt}'] \to [\X{rt}]
+   }
+
+
+.. index:: aggregate reference
+
+Aggregate Reference Instructions
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+.. _valid-struct.new:
+
+:math:`\STRUCTNEW~x`
+....................
+
+* The :ref:`defined type <syntax-deftype>` :math:`C.\CTYPES[x]` must exist.
+
+* The :ref:`expansion <aux-expand>` of :math:`C.\CTYPES[x]` must be a :ref:`structure type <syntax-structtype>` :math:`\TSTRUCT~\fieldtype^\ast`.
+
+* For each :ref:`field type <syntax-fieldtype>` :math:`\fieldtype_i` in :math:`\fieldtype^\ast`:
+
+  - Let :math:`\fieldtype_i` be :math:`\mut~\storagetype_i`.
+
+  - Let :math:`t_i` be the :ref:`value type <syntax-valtype>` :math:`\unpacktype(\storagetype_i)`.
+
+* Let :math:`t^\ast` be the concatenation of all :math:`t_i`.
+
+* Then the instruction is valid with type :math:`[t^\ast] \to [(\REF~x)]`.
+
+.. math::
+   \frac{
+     \expand(C.\CTYPES[x]) = \TSTRUCT~(\mut~\X{st})^\ast
+   }{
+     C \vdashinstr \STRUCTNEW~x : [(\unpacktype(\X{st}))^\ast] \to [(\REF~x)]
+   }
+
+.. _valid-struct.new_default:
+
+:math:`\STRUCTNEWDEFAULT~x`
+...........................
+
+* The :ref:`defined type <syntax-deftype>` :math:`C.\CTYPES[x]` must exist.
+
+* The :ref:`expansion <aux-expand>` of :math:`C.\CTYPES[x]` must be a :ref:`structure type <syntax-structtype>` :math:`\TSTRUCT~\fieldtype^\ast`.
+
+* For each :ref:`field type <syntax-fieldtype>` :math:`\fieldtype_i` in :math:`\fieldtype^\ast`:
+
+  - Let :math:`\fieldtype_i` be :math:`\mut~\storagetype_i`.
+
+  - Let :math:`t_i` be the :ref:`value type <syntax-valtype>` :math:`\unpacktype(\storagetype_i)`.
+
+  - The type :math:`t_i` must be :ref:`defaultable <valid-defaultable>`.
+
+* Let :math:`t^\ast` be the concatenation of all :math:`t_i`.
+
+* Then the instruction is valid with type :math:`[t^\ast] \to [(\REF~x)]`.
+
+.. math::
+   \frac{
+     \expand(C.\CTYPES[x]) = \TSTRUCT~(\mut~\X{st})^\ast
+     \qquad
+     (C \vdashvaltypedefaultable \unpacktype(\X{st}) \defaultable)^\ast
+   }{
+     C \vdashinstr \STRUCTNEWDEFAULT~x : [] \to [(\REF~x)]
+   }
+
+.. _valid-struct.get:
+.. _valid-struct.get_u:
+.. _valid-struct.get_s:
+
+:math:`\STRUCTGET\K{\_}\sx^?~x~i`
+.................................
+
+* The :ref:`defined type <syntax-deftype>` :math:`C.\CTYPES[x]` must exist.
+
+* The :ref:`expansion <aux-expand>` of :math:`C.\CTYPES[x]` must be a :ref:`structure type <syntax-structtype>` :math:`\TSTRUCT~\fieldtype^\ast`.
+
+* Let the :ref:`field type <syntax-fieldtype>` :math:`\mut~\storagetype` be :math:`\fieldtype^\ast[i]`.
+
+* Let :math:`t` be the :ref:`value type <syntax-valtype>` :math:`\unpacktype(\storagetype)`.
+
+* The extension :math:`\sx` must be present of and only if :math:`t` is different from :math:`\storagetype`.
```

Comment:
```suggestion
* The extension :math:`\sx` must be present if and only if :math:`t` is different from :math:`\storagetype`.
```
Why spec this condition as opposed to "if and only if :math:`\storagetype` is `i8` or `i16`"? The latter seems more straightforward.